pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,418 @@
+
+Note that by definition, the lexical scope of any subschema can be statically
+determined without taking
+[instances](https://json-schema.org/learn/glossary#instance) into account, just
+as we did here.
+
+### Lexical Scope and Anchors
+
+As another practical example, consider the `$anchor` keyword that defines a
+[location-independent
+identifier](https://json-schema.org/draft/2020-12/json-schema-core#section-8.2.2)
+for a schema. This keyword not only affects the schema object it is defined in,
+but its lexical scope. This is why declaring the same anchor identifier more
+than once in the same schema resource is an error (a clash in the lexical
+scope), while it is possible to declare the same anchor identifier on different
+schema resources (as the lexical scopes are different):
+
+![Example of anchors within and across lexical scopes](/img/posts/2024/understanding-lexical-dynamic-scopes/scopes-static-anchors.webp)
+
+### Following References
+
+When the evaluation process encounters a reference keyword, it *abandons* the
+lexical scope of the reference schema and enters the lexical scope of the
+*destination* schema.
+
+If the reference points to a subschema *within* the same schema resource, the
+lexical scope effectively remains the same. Coming back to the graph analogy,
+the evaluation process remains at the same node. However, if the reference
+points to a subschema on a *different* schema resource, the schema resource of
+the destination becomes the new lexical scope.  In the graph analogy, the
+evaluation process *follows an arrow* to another node.
+
+#### Within Schema Resources
+
+In the following example, the reference at `/items/$ref` points to
+`/$defs/person-name`. The destination schema is part of the same schema
+resource (the root schema resource), so the lexical scope remains the same:
+
+![Lexical scope after following a reference within the same resource](/img/posts/2024/understanding-lexical-dynamic-scopes/lexical-scope-ref-local.webp)
+
+#### Across Schema Resources
+
+Now consider the following sequence of examples. In the left, a JSON Schema
+called `https://example.com/point-in-time` with a nested schema resource (at
+`/$defs/timestamp`) and a reference to an external schema called
+`https://example.com/epoch` (from `/anyOf/1/$ref`). In the right, the
+corresponding directed graph representations of the root schema resource, the
+nested schema resource, and the external schema resource. Like before, at each
+step of the evaluation process, we gray out the parts of the schema and of the
+directed graph that are not part of the lexical scope.
+
+The evaluation process starts with the top level schema. The lexical scope at
+that point is the root schema resource, and both the nested schema resource and
+the external schema resource are out of scope: 
+
+![Lexical scope after following a reference accross resources (1)](/img/posts/2024/understanding-lexical-dynamic-scopes/lexical-scope-ref-remote-1.webp)
```

Comment:
The text shows it correctly, but the tree representation in the image seems to convey that the `anyOf` subschemas (the ones with the `$ref`s) aren't part of the top-level lexical scope, but they are.  I'm not sure how to convey that in the tree.

Probably not a big deal.

Reply:
I see your point. The labels in the arrow were meant to convey what reference is followed when following that arrow, and not whether they are part of the scope or not.

If you think its confusing, maybe worth getting rid of the arrow labels?

Reply:
Perhaps that.  The only other thing I can think of is to expand the dotted boundary to encompass the labels.  I'll let you decide.

Reply:
I removed the labels!